fawltydeps/packages.py

```diff
@@ -328,16 +336,57 @@ def temp_installed_requirements(requirements: List[str]) -> Iterator[Path]:
             (venv_dir / ".installed").touch()
             yield venv_dir
 
-    def lookup_packages(self, package_names: Set[str]) -> Dict[str, Package]:
+    # This function is only tagged as a context manager to facilitate an
+    # intechangeable use with `temp_installed_requirements()`
+    @staticmethod
+    @contextmanager
+    def cached_venv_with_requirements(
+        requirements: List[str], cache: pytest.Cache
+    ) -> Iterator[Path]:
+        """Install the given requirements into a cached venv.
+
+        Provide a path to the cached venv.
+        The venv persists after the call (and the installed deps are not
+        uninstalled).
+        The `--upgrade` and `--force-reinstall` options are not used to avoid
+        parallel tests over-writing the same package.
+        """
+
+        venv_dir = Path(
+            cache.mkdir(
+                f"fawltydeps_reused_venv_{sys.version_info.major}.{sys.version_info.minor}"
+            )
+        )
+        if not (venv_dir / ".installed").is_file():
+            venv.create(venv_dir, clear=True, with_pip=True)
+        subprocess.run(
+            [f"{venv_dir}/bin/pip", "install", "--no-deps"] + requirements,
+            check=True,  # fail if any of the commands fail
+        )
+        (venv_dir / ".installed").touch()
+        yield venv_dir
```

Comment:
This method is _very_ similar to the `temp_installed_requirements()` above, and I think some more refactoring can remove the duplication (while at the same time removing the pytest-specific stuff):

```python
    @staticmethod
    @contextmanager
    def installed_requirements(
        venv_dir: Path, requirements: List[str]
    ) -> Iterator[Path]:
        """Install the given requirements into venv_dir with `pip install`."""
        marker_file = venv_dir / ".installed"
        if not marker_file.is_file():
            venv.create(venv_dir, clear=True, with_pip=True)
        subprocess.run(
            [f"{venv_dir}/bin/pip", "install", "--no-deps"] + requirements,
            check=True,  # fail if any of the commands fail
        )
        marker_file.touch()
        yield venv_dir

    @classmethod
    @contextmanager
    def temp_installed_requirements(cls, requirements: List[str]) -> Iterator[Path]:
        """Create a temporary venv and install the given requirements into it.

        Provide a path to the temporary venv into the caller's context in which
        the given requirements have been `pip install`ed. Automatically remove
        the venv at the end of the context.
        """
        with tempfile.TemporaryDirectory() as tmpdir:
            with cls.installed_requirements(Path(tmpdir), requirements) as venv_dir:
                yield venv_dir
```
The first context manager takes care of creating a venv (only if needed) in a given `venv_dir` and installing the given requirements, and the second context manager then only needs to create a temporary directory, and call the first context manager to do the rest.

There is one difference in the argument list between the two context managers, but we can fix that with `functools.partial` below...